File: branches/objectweb/Foundation/orbsupport/java/src/cdmw/orbsupport/ORBSupport.java
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES 
 * and SELEX-SI. 
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003. 
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under 
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your 
 * option) any later version. 
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public 
 * License for more details. 
 * 
 * You should have received a copy of the GNU Library General 
 * Public License along with CARDAMOM; see the file COPYING. If not, write to 
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


package cdmw.orbsupport;

import cdmw.common.Options;
import cdmw.ossupport.OS;

import org.omg.CORBA.CompletionStatus;
import org.omg.CORBA.SystemException;
import org.omg.CORBA.BAD_PARAM;

import org.omg.PortableServer.POA;
import org.omg.PortableServer.POAManager;
import org.omg.PortableServer.POAPackage.AdapterAlreadyExists;
import org.omg.PortableServer.POAPackage.InvalidPolicy;
import java.lang.*;

/**
 * This class provides a wrapping of the ORB dependent capability.
 *
 * Features: Thread safe
 *
 */
public class ORBSupport {

    /**
     * Name of the CDMW POA root of all CDMW-specific POAs
     */
    public final static String CDMW_ROOT_POA_NAME = "CDMW_RootPOA";

    /**
     * Minimum port number.
     */
    public final static long MIN_PORT_NUMBER = 0;

    /**
     * Maximum port number.
     */
    public final static long MAX_PORT_NUMBER = 65535;

    /**
     * Implementation of the interface ORBSupportInterface
     */
    private final static ORBSupportInterface orbSupport = new ORBSupportImpl();
    

    /**
     * Creates a new ORB instance, initialised with the
     * specified CDMW StrategyList. This method calls the
     * underlying ORB specific ORBInit, so the semantic is as
     * defined by CORBA.
     *
     * @param args The command line arguments
     * @param strategyList A list of CDMW strategies
     * @param orbIdentifier Optional ORB identifier
     *
     * @return A valid reference to an ORB pseudo-object
     *
     * @exception BAD_PARAM Invalid port number.
     */
    public static org.omg.CORBA.ORB ORBInit(
        String[] args, 
        StrategyList strategyList,
        String orbIdentifier) 
        throws SystemException {
		
	int minor;
        String rootPOAPortOption = OS.getOptionValue(args,
        Options.LOCALISATION_SERVICE_OPTION);
        int rootPOAPortNumber = 0;
        if ( !rootPOAPortOption.equals("no") ) {
            if ( !rootPOAPortOption.equals("yes") ) {
                try {
                    int port = Integer.valueOf(rootPOAPortOption).intValue();
                    if ( port > MIN_PORT_NUMBER && port < MAX_PORT_NUMBER ) {
                        rootPOAPortNumber = port;
                    } else {
                        minor = ExceptionMinorCodes.BAD_PARAMInvalidPortNumber;
                        throw new BAD_PARAM(
                            ExceptionStrings.BAD_PARAM_reasons[minor],
                            minor, CompletionStatus.COMPLETED_NO);
                    }
                }
                catch(NumberFormatException nfe) {
                    minor = ExceptionMinorCodes.BAD_PARAMInvalidPortNumber;
                    throw new BAD_PARAM(
                        ExceptionStrings.BAD_PARAM_reasons[minor],
                        minor, CompletionStatus.COMPLETED_NO);
                }
            } else {
                minor = ExceptionMinorCodes.BAD_PARAMInvalidPortNumber;
                throw new BAD_PARAM(
                    ExceptionStrings.BAD_PARAM_reasons[minor],
                    minor, CompletionStatus.COMPLETED_NO);
            }
	 
        } // else ignore

        String cdmwRootPOAPortOption = OS.getOptionValue(args,
            Options.PROCESS_PORT_OPTION);
        int cdmwRootPOAPortNumber = 0;
        if ( !cdmwRootPOAPortOption.equals("no") ) {
            minor = ExceptionMinorCodes.BAD_PARAMInvalidPortNumber;
            if ( !cdmwRootPOAPortOption.equals("yes") ) {
                try {
                    int port =
                        Integer.valueOf(cdmwRootPOAPortOption).intValue();
                    if ( port > MIN_PORT_NUMBER && port < MAX_PORT_NUMBER ) {
                        cdmwRootPOAPortNumber = port;
                    } else {
                        minor = ExceptionMinorCodes.BAD_PARAMInvalidPortNumber;
                        throw new BAD_PARAM(
                            ExceptionStrings.BAD_PARAM_reasons[minor],
                            minor, CompletionStatus.COMPLETED_NO);
                    }
                }
                catch(NumberFormatException nfe) {
                    minor = ExceptionMinorCodes.BAD_PARAMInvalidPortNumber;
                    throw new BAD_PARAM(
                        ExceptionStrings.BAD_PARAM_reasons[minor],
                        minor, CompletionStatus.COMPLETED_NO);
                }
            } else {
                minor = ExceptionMinorCodes.BAD_PARAMInvalidPortNumber;
                throw new BAD_PARAM(
                    ExceptionStrings.BAD_PARAM_reasons[minor],
                    minor, CompletionStatus.COMPLETED_NO);
            }
        } // else ignore


        return ORBInit(args, strategyList, cdmwRootPOAPortNumber,
                       rootPOAPortNumber, orbIdentifier);

    }
    
    /**
     * Calls the underlying ORB specific ORBInit, so the semantic is as
     * defined by CORBA.
     *
     * @param args The command line arguments
     * @param strategyList A list of CDMW strategies
     *
     * @return A valid reference to an ORB pseudo-object
     *
     * @exception SystemException Any CORBA system exception that
     * may be raised by the underlying ORB.
     */
    public static org.omg.CORBA.ORB ORBInit(
        String[] args, StrategyList strategyList) throws SystemException {
        return ORBInit(args, strategyList, "");
    }
    
    /**
     * Calls the underlying ORB specific ORBInit, so the semantic is as
     * defined by CORBA.
     *
     * @param args The command line arguments
     * @param strategyList A list of CDMW strategies
     * @param cdmwRootPOAPort specific port for the Cdmw root POA
     * @param rootPOAPort specific port for the root POA
     * @param orbIdentifier Optional ORB identifier
     *
     * @return A valid reference to an ORB pseudo-object
     *
     * @exception SystemException Any CORBA system exception that
     * may be raised by the underlying ORB.
     */
    public static org.omg.CORBA.ORB ORBInit(
        String[] args,
        StrategyList strategyList,
        int cdmwRootPOAPort,
        int rootPOAPort,
        String orbIdentifier)
        throws org.omg.CORBA.SystemException {
    
        return orbSupport.ORBInit(
            args, 
            strategyList, 
            cdmwRootPOAPort, 
            rootPOAPort,
            orbIdentifier);
            
    }

    /**
     * Creates a POA child of the specified parentPOA. The
     *  created POA is initialised with the specified policies
     *  and with others Policy, created according to the
     *  specified CDMW StrategyList.
     *
     * @param parentPOA The parent POA of the POA to be created
     * @param adapterName Name of the new POA
     * @param aPOAManager Manager of the POA
     * @param policies List of policies to use to create the POA
     * @param strategyList List of CDMW Strategies
     *
     * @return A reference to the created POA
     *
     * @exception SystemException Any CORBA system exception that
     * may be raised by the underlying ORB.
     * @exception AdapterAlreadyExists A POA with the same name
     * is already created by parentPOA.
     * @exception InvalidPolicy Supplied policies are not compatible.
     *
     * @see StrategyList
     */
    public static POA createPOA(POA parentPOA,
        String adapterName, POAManager aPOAManager,
        PolicyList policies,StrategyList strategyList)
        throws SystemException, AdapterAlreadyExists, InvalidPolicy {

        return orbSupport.createPOA(parentPOA,adapterName,aPOAManager,
                                    policies,strategyList);

    }
    
    /**
     * Creates a POA child of the specified parentPOA. The
     *  created POA is initialised with the specified policies
     *  and with others Policy, created according to the
     *  specified CDMW StrategyList.
     *
     * @param parentPOA The parent POA of the POA to be created
     * @param adapterName Name of the new POA
     * @param aPOAManager Manager of the POA
     * @param policies List of policies to use to create the POA
     * @param strategyList List of CDMW Strategies
     *
     * @return A reference to the created POA
     *
     * @exception SystemException Any CORBA system exception that
     * may be raised by the underlying ORB.
     * @exception AdapterAlreadyExists A POA with the same name
     * is already created by parentPOA.
     * @exception InvalidPolicy Supplied policies are not compatible.
     *
     * @see StrategyList
     */
    public static POA createPOA(POA parentPOA,
        String adapterName, POAManager aPOAManager,
        org.omg.CORBA.Policy[] policies,StrategyList strategyList)
        throws SystemException, AdapterAlreadyExists, InvalidPolicy {

        return orbSupport.createPOA(parentPOA,adapterName,aPOAManager,
                                    new PolicyList(policies),strategyList);

    }

    /**
     * Binds the specified object reference
     * to a name reachable by corbaloc URLs.
     *
     * @param orb The ORB pseudo-object
     * @param corbalocName Name to be assigned to reach the object
     * @param o Object reference to bind to the corbalocName
     *
     * @exception SystemException Any CORBA system exception that
     * may be raised by the underlying ORB.
     */
    public static void bindObjectToCorbaloc(
        org.omg.CORBA.ORB orb, 
        String corbalocName,
        org.omg.CORBA.Object o) 
        throws SystemException {
        orbSupport.bindObjectToCorbaloc(orb,corbalocName,o);
    }

    /**
     * Removes a corbaloc binding.
     *
     * @param orb The ORB pseudo-object
     * @param corbalocName Name of the binding to be removed.
     *
     * @exception SystemException Any CORBA system exception that
     * may be raised by the underlying ORB.
     */
    public static void removeCorbalocBinding(
        org.omg.CORBA.ORB orb, 
        String corbalocName)
        throws SystemException {
        orbSupport.removeCorbalocBinding(orb,corbalocName);
    }
    
    /**
     * Register the object reference as specified service 
     * to be reachable with "resolveInitialReference" API.
     * 
     * @param orb The ORB pseudo-object
     * @param serviceName Name to be assigned as service to reach the object
     * @param object Object reference to associate to service name
     *
     * @exception BAD_PARAM Invalid parameter. The CDMW
     * <B>BAD_PARAMCorbalocBindingAlreadyExists</B> minor code indicates that
     * an object reference has already been bound to the specified name.
     */ 
    public static void registerInitialReference(
        org.omg.CORBA.ORB orb,
        String serviceName, 
        org.omg.CORBA.Object object) 
        throws SystemException {
        orbSupport.registerInitialReference(orb,serviceName,object);
    }
    
    /**
     * Translates an Exception into a string in
     * a human readable format.
     * 
     * @param e The CORBA exception in question
     * @return A string in a human readable format
     */ 
    public static String exceptionToString(Exception e) {
        
        if (e instanceof org.omg.CORBA.SystemException) {
            return systemExceptionToString((org.omg.CORBA.SystemException) e);
        } else {
            return e.toString();
        }
        
    }

    /**
     * Translates an org.omg.CORBA.SystemException into a string 
     * in a human readable format.
     * 
     * @param ex The CORBA system exception in question
     * @return A string in a human readable format
     */ 
    private static String 
        systemExceptionToString(org.omg.CORBA.SystemException se) {
        
        if ((se.minor & ExceptionMinorCodes.VMCID_MASK) 
            == ExceptionMinorCodes.CDMW_VMCID) {
            
            int minor = se.minor & ExceptionMinorCodes.MINOR_CODE_MASK;

            // This is a CDMW system exception.
            if (se instanceof org.omg.CORBA.BAD_INV_ORDER) {
                return ExceptionStrings.BAD_INV_ORDER_reasons[minor];
            } else if(se instanceof org.omg.CORBA.BAD_PARAM) {
                return ExceptionStrings.BAD_INV_ORDER_reasons[minor];
            } else if(se instanceof org.omg.CORBA.INTERNAL) {
                return ExceptionStrings.BAD_PARAM_reasons[minor];
            } else if(se instanceof org.omg.CORBA.NO_IMPLEMENT) {
                return ExceptionStrings.NO_IMPLEMENT_reasons[minor];
            } else if(se instanceof org.omg.CORBA.NO_MEMORY) {
                return ExceptionStrings.NO_MEMORY_reasons[minor];
            } else if(se instanceof org.omg.CORBA.NO_PERMISSION) {
                return ExceptionStrings.NO_PERMISSION_reasons[minor];
            } else if(se instanceof org.omg.CORBA.NO_RESOURCES) {
                return ExceptionStrings.NO_RESOURCES_reasons[minor];
            } else if(se instanceof org.omg.CORBA.OBJECT_NOT_EXIST) {
                return ExceptionStrings.OBJECT_NOT_EXIST_reasons[minor];
            } else if(se instanceof org.omg.CORBA.TRANSIENT) {
                return ExceptionStrings.TRANSIENT_reasons[minor];
            } else {
                return "";
            }

        } else {
            return se.toString();
        }
        
    }
    
    /**
     * Clean up the orb
     *
     * @param orb The ORB pseudo-object
     * @exception BAD_INV_ORDER 
     */
    public static void ORBCleanup(org.omg.CORBA.ORB orb) 
        throws org.omg.CORBA.BAD_INV_ORDER {
        orbSupport.ORBCleanup(orb);
    }

}

File: Foundation/orbsupport/c++/src/OrbSupport.cpp
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES 
 * and SELEX-SI. 
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003. 
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under 
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your 
 * option) any later version. 
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public 
 * License for more details. 
 * 
 * You should have received a copy of the GNU Library General 
 * Public License along with CARDAMOM; see the file COPYING. If not, write to 
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


#include "Foundation/orbsupport/CORBA.hpp"
#include "Foundation/orbsupport/OrbSupport.hpp"
#include "Foundation/orbsupport/ExceptionMinorCodes.hpp"
#include "Foundation/common/Options.hpp"
#include "Foundation/ossupport/OS.hpp"
#include "orbsupport/ExceptionStrings.hpp"

#if CDMW_ORB_VDR == orbacus

#include "orbsupport/OrbOrbacusImpl.hpp"

#elif CDMW_ORB_VDR == tao

#include "orbsupport/OrbTaoImpl.hpp"

#endif

#include <sstream>
#include <string>

#define MAKE_EXCEPTION_TO_STRING(T) \
    std::string T##_to_string(const CORBA::##T & ex) \
    throw (Cdmw::OutOfMemoryException)  \
    { \
        return system_exception_string(#T, \
                                       ex.minor(), \
                                       ex.completed(), \
                                       Cdmw::OrbSupport::ExceptionStrings::##T##_reasons); \
    }

/*
#define CHECK_AND_RETURN_SYSTEM_EXCEPTION(T) \
            (ex_name == #T) \
            do { \
                const CORBA::##T * sys_ex \
                    = CORBA::##T::_downcast(&ex); \
                return T##_to_string(*sys_ex); \
            } while(0)
*/ 
// -----
#define CHECK_AND_RETURN_SYSTEM_EXCEPTION(T) \
    (ex_name == #T) \
    do { \
        CORBA::##T * sys_ex \
        = CORBA::##T::_downcast( (CORBA::Exception*)&ex ); \
        return T##_to_string(*sys_ex); \
    } while(0)
// -----



#if HAS_CORBA_EXCEPTION_OSTREAM_OPERATOR_DEFINITION == 0

//
// Exception output on std::ostream
//
std::ostream &operator<<(std::ostream& os, const CORBA::Exception& ex)
{
    try
    {
        os << Cdmw::OrbSupport::OrbSupport::exception_to_string(ex);
    }
    catch (...)
    {
        // ignore ...
    }

    return os;
}

#endif

namespace
{
    // VMCID (Vendor Minor Codeset ID) mask
    const CORBA::ULong VMCID_MASK = 0xfffff000;
    const CORBA::ULong MINOR_CODE_MASK = 0x00000fff;

    const long MIN_PORT_NUMBER = 0;
    const long MAX_PORT_NUMBER = 65535;

    std::string system_exception_string(const std::string & exception_name,
                                        CORBA::ULong minor_code,
                                        CORBA::CompletionStatus completion_status,
                                        const char* messages[])
    throw(Cdmw::OutOfMemoryException)
    {
        try
        {
            std::ostringstream s;
            s << "System exception `" << exception_name << "'" << "\n";
            s << "Completed: ";

            switch (completion_status)
            {

                    case CORBA::COMPLETED_YES:
                    s << "yes";
                    break;

                    case CORBA::COMPLETED_NO:
                    s << "no";
                    break;

                    case CORBA::COMPLETED_MAYBE:
                    s << "maybe";
                    break;
            }

            s << "\n";
            s << "Minor code: " << std::hex
            << (unsigned long)minor_code << std::dec << "\n";

            s << "(" << messages[minor_code & MINOR_CODE_MASK] << ")";

            return s.str();
        }
        catch (const std::bad_alloc & )
        {
            CDMW_THROW(Cdmw::OutOfMemoryException);
        }
    }

    using namespace CORBA;

    std::string BAD_INV_ORDER_to_string(const CORBA::BAD_INV_ORDER & ex)
    throw (Cdmw::OutOfMemoryException)
    {
        return system_exception_string("BAD_INV_ORDER",
                                       ex.minor(),
                                       ex.completed(),
                                       Cdmw::OrbSupport::ExceptionStrings::BAD_INV_ORDER_reasons);
    }

    std::string BAD_PARAM_to_string(const CORBA::BAD_PARAM & ex)
    throw (Cdmw::OutOfMemoryException)
    {
        return system_exception_string("BAD_PARAM",
                                       ex.minor(),
                                       ex.completed(),
                                       Cdmw::OrbSupport::ExceptionStrings::BAD_PARAM_reasons);
    }

    std::string INTERNAL_to_string(const CORBA::INTERNAL & ex)
    throw (Cdmw::OutOfMemoryException)
    {
        return system_exception_string("INTERNAL",
                                       ex.minor(),
                                       ex.completed(),
                                       Cdmw::OrbSupport::ExceptionStrings::INTERNAL_reasons);
    }

    std::string NO_IMPLEMENT_to_string(const CORBA::NO_IMPLEMENT & ex)
    throw (Cdmw::OutOfMemoryException)
    {
        return system_exception_string("NO_IMPLEMENT",
                                       ex.minor(),
                                       ex.completed(),
                                       Cdmw::OrbSupport::ExceptionStrings::NO_IMPLEMENT_reasons);
    }

    std::string NO_MEMORY_to_string(const CORBA::NO_MEMORY & ex)
    throw (Cdmw::OutOfMemoryException)
    {
        return system_exception_string("NO_MEMORY",
                                       ex.minor(),
                                       ex.completed(),
                                       Cdmw::OrbSupport::ExceptionStrings::NO_MEMORY_reasons);
    }

    std::string NO_PERMISSION_to_string(const CORBA::NO_PERMISSION & ex)
    throw (Cdmw::OutOfMemoryException)
    {
        return system_exception_string("NO_PERMISSION",
                                       ex.minor(),
                                       ex.completed(),
                                       Cdmw::OrbSupport::ExceptionStrings::NO_PERMISSION_reasons);
    }

    std::string NO_RESOURCES_to_string(const CORBA:: NO_RESOURCES& ex)
    throw (Cdmw::OutOfMemoryException)
    {
        return system_exception_string("NO_RESOURCES",
                                       ex.minor(),
                                       ex.completed(),
                                       Cdmw::OrbSupport::ExceptionStrings::NO_RESOURCES_reasons);
    }

    std::string OBJECT_NOT_EXIST_to_string(const CORBA:: OBJECT_NOT_EXIST& ex)
    throw (Cdmw::OutOfMemoryException)
    {
        return system_exception_string("OBJECT_NOT_EXIST",
                                       ex.minor(),
                                       ex.completed(),
                                       Cdmw::OrbSupport::ExceptionStrings::OBJECT_NOT_EXIST_reasons);
    }

    std::string TRANSIENT_to_string(const CORBA::TRANSIENT & ex)
    throw (Cdmw::OutOfMemoryException)
    {
        return system_exception_string("TRANSIENT",
                                       ex.minor(),
                                       ex.completed(),
                                       Cdmw::OrbSupport::ExceptionStrings::TRANSIENT_reasons);
    }



    //   MAKE_EXCEPTION_TO_STRING(BAD_INV_ORDER)
    //     MAKE_EXCEPTION_TO_STRING(BAD_PARAM)
    //     MAKE_EXCEPTION_TO_STRING(INTERNAL)
    //     MAKE_EXCEPTION_TO_STRING(NO_IMPLEMENT)
    //     MAKE_EXCEPTION_TO_STRING(NO_MEMORY)
    //     MAKE_EXCEPTION_TO_STRING(NO_PERMISSION)
    //     MAKE_EXCEPTION_TO_STRING(NO_RESOURCES)
    //     MAKE_EXCEPTION_TO_STRING(OBJECT_NOT_EXIST)
    //     MAKE_EXCEPTION_TO_STRING(TRANSIENT)
}

; // End anonymous namespace

namespace Cdmw
{

    namespace OrbSupport
    {

        const char* CDMW_ROOT_POA_NAME = "CDMW_RootPOA";

#if CDMW_ORB_VDR == orbacus

        typedef OrbOrbacusImpl UnderlyingORB;

#elif CDMW_ORB_VDR == tao

        typedef OrbTaoImpl UnderlyingORB;

#endif

        /**
        *Implementation notes: none 
        *<p>
        *Portability issues: none
        *<p>
        */

        //
        // Returns an initiliased ORB
        //
        CORBA::ORB_ptr
        OrbSupport::ORB_init (int& argc, char** argv,
                              const StrategyList& strategylist,
                              const char* orb_identifier)
        throw (CORBA::SystemException)
        {
            // Is CDMW Process port number specified?
            // Check Options::PROCESS_PORT_OPTION in argv for the value of the port number

            using namespace Cdmw::OsSupport;

            using namespace Cdmw::Common;
            std::string root_poa_port_option;

            try
            {
                root_poa_port_option = OS::get_option_value(argc, argv,
                                       Options::LOCALISATION_SERVICE_OPTION);
            }
            catch (const OutOfMemoryException &)
            {
                throw CORBA::NO_MEMORY(NO_MEMORY, CORBA::COMPLETED_NO);
            }
            catch (const InternalErrorException &)
            {
                throw CORBA::INTERNAL(INTERNAL, CORBA::COMPLETED_NO);
            }


            std::size_t root_poa_port_number(0);

            if (root_poa_port_option != "no")
            {
                if (root_poa_port_option != "yes")
                {
                    long port;
                    std::istringstream strPoaPort(root_poa_port_option);
                    strPoaPort >> port;

                    if (strPoaPort.fail() ||
                            !(port > MIN_PORT_NUMBER && port < MAX_PORT_NUMBER))
                    {
                        throw CORBA::BAD_PARAM(BAD_PARAMInvalidPortNumber, CORBA::COMPLETED_NO);
                    }
                    else
                        root_poa_port_number = port;
                }
                else
                {
                    throw CORBA::BAD_PARAM(BAD_PARAMInvalidPortNumber, CORBA::COMPLETED_NO);
                }
            } // else ignore.

            std::string cdmw_root_poa_port_option;

            try
            {
                cdmw_root_poa_port_option = OS::get_option_value (argc, argv,
                                             Options::PROCESS_PORT_OPTION);
            }
            catch (const OutOfMemoryException &)
            {
                throw CORBA::NO_MEMORY(NO_MEMORY, CORBA::COMPLETED_NO);
            }
            catch (const InternalErrorException &)
            {
                throw CORBA::INTERNAL(INTERNAL, CORBA::COMPLETED_NO);
            }


            std::size_t cdmw_root_poa_port_number(0);

            if (cdmw_root_poa_port_option != "no")
            {
                if (cdmw_root_poa_port_option != "yes")
                {
                    long port;
                    std::istringstream strPoaPort(cdmw_root_poa_port_option);
                    strPoaPort >> port;

                    if (strPoaPort.fail() ||
                            !(port > MIN_PORT_NUMBER && port < MAX_PORT_NUMBER))
                    {
                        throw CORBA::BAD_PARAM(BAD_PARAMInvalidPortNumber, CORBA::COMPLETED_NO);
                    }
                    else
                        cdmw_root_poa_port_number = port;
                }
                else
                {
                    throw CORBA::BAD_PARAM(BAD_PARAMInvalidPortNumber, CORBA::COMPLETED_NO);
                }
            } // else ignore.




            return UnderlyingORB::ORB_init (argc, argv, strategylist,
                                            root_poa_port_number,
                                            cdmw_root_poa_port_number,
                                            orb_identifier);
        }


        //
        // Create a POA, child of the specified POA, with the right
        //    policies to respect the Cdmw Strategies.
        //
        PortableServer::POA_ptr
        OrbSupport::create_POA (PortableServer::POA_ptr parentPOA,
                                const char* adapter_name,
                                PortableServer::POAManager_ptr a_POAManager,
                                const CORBA::PolicyList& policies,
                                const StrategyList& strategyList)
        throw (PortableServer::POA::AdapterAlreadyExists,
               PortableServer::POA::InvalidPolicy,
               CORBA::SystemException)
        {
            return UnderlyingORB::create_POA(parentPOA, adapter_name, a_POAManager, policies,
                                             strategyList);
        }



        //
        // Make the current thread adopted by the underlying thread
        //    system
        //
        ThreadAdoptionToken*
        OrbSupport::adopt_thread()
        throw(OutOfMemoryException, InternalErrorException)
        {
#if CDMW_ORB_VDR == orbacus

            return UnderlyingORB::adopt_thread();

#else

            return NULL;

#endif
        }



        //
        // Transform a CORBA excpetion in a human readable format
        //
        std::string
        OrbSupport::exception_to_string(const CORBA::Exception& ex )
        throw(OutOfMemoryException)
        {
            /*
            // Is it a system exception?
            const CORBA::SystemException* sys 
            = CORBA::SystemException::_downcast(&ex);
            */

            // -----
            // Is it a system exception?
            const CORBA::SystemException* sys
            = CORBA::SystemException::_downcast( (CORBA::Exception*) & ex );
            // -----

            if (sys)
                return system_exception_to_string(*sys);
            else
                return UnderlyingORB::exception_to_string( ex );
        }

        std::string
        OrbSupport::system_exception_to_string(const CORBA::SystemException& ex)
        throw(OutOfMemoryException)
        {
            try
            {
                CORBA::ULong minor_code = ex.minor();

                if ((minor_code & VMCID_MASK) == CDMW_VMCID)
                {
                    // This is a CDMW system exception.
                    std::string ex_name = UnderlyingORB::exception_name(ex);
                    //             if CHECK_AND_RETURN_SYSTEM_EXCEPTION(BAD_INV_ORDER);
                    //             else if CHECK_AND_RETURN_SYSTEM_EXCEPTION(BAD_PARAM);
                    //             else if CHECK_AND_RETURN_SYSTEM_EXCEPTION(INTERNAL);
                    //             else if CHECK_AND_RETURN_SYSTEM_EXCEPTION(NO_IMPLEMENT);
                    //             else if CHECK_AND_RETURN_SYSTEM_EXCEPTION(NO_MEMORY);
                    //             else if CHECK_AND_RETURN_SYSTEM_EXCEPTION(NO_PERMISSION);
                    //             else if CHECK_AND_RETURN_SYSTEM_EXCEPTION(NO_RESOURCES);
                    //             else if CHECK_AND_RETURN_SYSTEM_EXCEPTION(OBJECT_NOT_EXIST);
                    //             else if CHECK_AND_RETURN_SYSTEM_EXCEPTION(TRANSIENT);
                    //             else
                    //                 return "";


                    if (ex_name == "BAD_INV_ORDER")
                        do
                        {
                            CORBA::BAD_INV_ORDER * sys_ex \
                            = CORBA::BAD_INV_ORDER::_downcast( (CORBA::Exception*) & ex );
                            return BAD_INV_ORDER_to_string(*sys_ex);
                        }
                        while (0);
                    else if (ex_name == "BAD_PARAM")
                        do
                        {
                            CORBA::BAD_PARAM * sys_ex
                            = CORBA::BAD_PARAM::_downcast( (CORBA::Exception*) & ex );
                            return BAD_PARAM_to_string(*sys_ex);
                        }
                        while (0);
                    else if (ex_name == "INTERNAL")
                        do
                        {
                            CORBA::INTERNAL * sys_ex
                            = CORBA::INTERNAL::_downcast( (CORBA::Exception*) & ex );
                            return INTERNAL_to_string(*sys_ex);
                        }
                        while (0);
                    else if (ex_name == "NO_IMPLEMENT")
                        do
                        {
                            CORBA::NO_IMPLEMENT * sys_ex
                            = CORBA::NO_IMPLEMENT::_downcast( (CORBA::Exception*) & ex );
                            return NO_IMPLEMENT_to_string(*sys_ex);
                        }
                        while (0);
                    else if (ex_name == "NO_MEMORY")
                        do
                        {
                            CORBA::NO_MEMORY * sys_ex
                            = CORBA::NO_MEMORY::_downcast( (CORBA::Exception*) & ex );
                            return NO_MEMORY_to_string(*sys_ex);
                        }
                        while (0);
                    else if (ex_name == "NO_PERMISSION")
                        do
                        {
                            CORBA::NO_PERMISSION * sys_ex
                            = CORBA::NO_PERMISSION::_downcast( (CORBA::Exception*) & ex );
                            return NO_PERMISSION_to_string(*sys_ex);
                        }
                        while (0);
                    else if (ex_name == "NO_RESOURCES")
                        do
                        {
                            CORBA::NO_RESOURCES * sys_ex
                            = CORBA::NO_RESOURCES::_downcast( (CORBA::Exception*) & ex );
                            return NO_RESOURCES_to_string(*sys_ex);
                        }
                        while (0);
                    else if (ex_name == "OBJECT_NOT_EXIST")
                        do
                        {
                            CORBA::OBJECT_NOT_EXIST * sys_ex
                            = CORBA::OBJECT_NOT_EXIST::_downcast( (CORBA::Exception*) & ex );
                            return OBJECT_NOT_EXIST_to_string(*sys_ex);
                        }
                        while (0);
                    else if (ex_name == "TRANSIENT")
                        do
                        {
                            CORBA::TRANSIENT * sys_ex
                            = CORBA::TRANSIENT::_downcast( (CORBA::Exception*) & ex );
                            return TRANSIENT_to_string(*sys_ex);
                        }
                        while (0);
                    else
                        return "";


                }
                else
                    return UnderlyingORB::exception_to_string( ex );
            }
            catch (const OutOfMemoryException & )
            {
                throw;
            }
            catch (const std::bad_alloc & )
            {
                CDMW_THROW(OutOfMemoryException);
            }
        }



        void
        OrbSupport::bind_object_to_corbaloc (CORBA::ORB_ptr orb,
                                             const std::string& corbaloc_name,
                                             CORBA::Object_ptr object)
        throw (CORBA::SystemException)
        {
            UnderlyingORB::bind_object_to_corbaloc (orb, corbaloc_name, object);
        }

        void
        OrbSupport::remove_corbaloc_binding (CORBA::ORB_ptr orb,
                                             const std::string& corbaloc_name)
        throw (CORBA::SystemException)
        {
            UnderlyingORB::remove_corbaloc_binding (orb, corbaloc_name);
        }


        void
        OrbSupport::register_initial_reference (CORBA::ORB_ptr orb,
                                                const std::string& service_name,
                                                CORBA::Object_ptr object)
        throw (CORBA::SystemException)
        {
            UnderlyingORB::register_initial_reference (orb, service_name, object);
        }


        void
        OrbSupport::ORB_cleanup(CORBA::ORB_ptr orb)
        throw(CORBA::BAD_INV_ORDER)
        {
            UnderlyingORB::ORB_cleanup(orb);
        }
        
        void                
        OrbSupport::set_endpoint_selector_factory (const std::string& endpoint_selector_factory)
        {
            UnderlyingORB::set_endpoint_selector_factory(endpoint_selector_factory);
        }
        
        
        std::string
        OrbSupport::get_repository_id_from_ior(CORBA::ORB_ptr orb, CORBA::Object_ptr object)
            throw( CORBA::SystemException )
        {
            return UnderlyingORB::get_repository_id_from_ior(orb, object);
        }
        
    } // End namespace OrbSupport
} // End namespace Cdmw

